api/search.rb

```diff
@@ -1,58 +1,60 @@
 def get_thread_ids(context, group_ids, local_params, search_text)
-  filters = []
-  filters.push({term: {commentable_id: local_params['commentable_id']}}) if local_params['commentable_id']
-  filters.push({terms: {commentable_id: local_params['commentable_ids'].split(',')}}) if local_params['commentable_ids']
-  filters.push({term: {course_id: local_params['course_id']}}) if local_params['course_id']
+  must = []
+  filter = []
+  must.push({term: {commentable_id: local_params['commentable_id']}}) if local_params['commentable_id']
+  must.push({terms: {commentable_id: local_params['commentable_ids'].split(',')}}) if local_params['commentable_ids']
+  must.push({term: {course_id: local_params['course_id']}}) if local_params['course_id']
+  must.push(
+    {
+      multi_match: {
+        query: search_text,
+        fields: [:title, :body],
+        operator: :AND
+      }
+    }
+  )
+  group_id = local_params['group_id']
+
+  if group_id
+    filter.push(
+        {:bool => {:must_not => {:exists => {:field => :group_id}}}},
+        {:term => {:group_id => group_id}}
+    )
+  end
 
-  filters.push({or: [
-      {not: {exists: {field: :context}}},
-      {term: {context: context}}
-  ]})
+  filter.push(
+      {:bool => {:must_not => {:exists => {:field => context}}}},
+      {:term => {:context => context}}
+  )
 
   unless group_ids.empty?
-    filters.push(
-        {
-            bool: {
-                should: [
-                    {:not => {:exists => {:field => :group_id}}},
-                    {:terms => {:group_id => group_ids}}
-                ]
-            }
-        }
+    filter.push(
+        {:bool => {:must_not => {:exists => {:field => :group_id}}}},
+        {:terms => {:group_id => group_ids}}
     )
   end
 
   body = {
-      size: CommentService.config['max_deep_search_comment_count'].to_i,
-      sort: [
-          {updated_at: :desc}
-      ],
-      query: {
-          filtered: {
-              query: {
-                  multi_match: {
-                      query: search_text,
-                      fields: [:title, :body],
-                      operator: :AND
-                  }
-              },
-              filter: {
-                  bool: {
-                      must: filters
-                  }
-              }
-          }
+    size: CommentService.config['max_deep_search_comment_count'].to_i,
+    sort: [
+      {updated_at: :desc}
+    ],
+    query: {
+      bool: {
+        must: must,
```

Comment:
Has this changed the search behavior, splitting the `filters` from the old code into `must` and `should`? From line 41 on the other side, I would have expected they were all "musts" before.

Reply:
Maybe I'll just generalize that question to ask - have the changes to this file resulted in any differences in how we perform our queries/what results they return?  Can you tell me a little about what needed to change here?  It's hard to follow from the diff alone.

Reply:
As I see from test results we get expected result of search behaviour. 
The filtered query is replaced by the bool query. Instead of the following. I use should to optimise filter query which look like, for example: 
```
filter: {
                  bool: {
                      must: filters
                  }
              }
```
where `filters` 
```
{
            bool: {
                should: [
                    {:not => {:exists => {:field => :group_id}}},
                    {:terms => {:group_id => group_ids}}
                ]
            }
        }
```
And ES haven't critical difference between `should` and `filter` clauses. [Useful doc link](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-bool-query.html#query-dsl-bool-query) which explain `should` and `filter` difference. 
Now search query looks simpler, I think.